### PR TITLE
Track unclaimed tasks in Redis.

### PR DESCRIPTION
**Pull request by vadimberezniker**
Sharing the state across schedulers will give us better scheduling
during upscaling.

We use a sorted set per executor pool to track unclaimed tasks.

---

**Version bump**: Patch

### CODE REVIEW

**Comment by vadimberezniker** [comment]
Fixed the diffbase, PTAL.

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go [hunk 1]
```diff
@@ -361,70 +367,39 @@ func (np *nodePool) FindConnectedExecutorByID(executorID string) *executionNode
...
+	// The next 2 commands are not atomic but it's okay if the list length is not exactly what we want.
+	n, err := np.rdb.ZCard(ctx, key).Result()
+	if err != nil {
+		return err
 	}
-	delete(l.taskMap, taskID)
-	l.taskList.Remove(e)
+	if n > maxUnclaimedTasksTracked {
+		return np.rdb.ZRemRangeByRank(ctx, key, 0, n-maxUnclaimedTasksTracked-1).Err()
+	}
```
Comment (tylerwilliams) [inline enterprise/server/scheduling/scheduler_server/scheduler_server.go:393]: lol not afraid to admit i had to think about this for way too long to figure it out. Could you add a comment that explains what ZREMRANGEBYRANK is doing and also explains the n-maxTasks-1 math?

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go [hunk 2]
```diff
@@ -361,70 +367,39 @@ func (np *nodePool) FindConnectedExecutorByID(executorID string) *executionNode
...
+}
 
-	// Random sample (without replacement) up to `count` tasks from the
-	// returned results.
-	rand.Shuffle(len(unclaimed), func(i, j int) {
-		unclaimed[i], unclaimed[j] = unclaimed[j], unclaimed[i]
-	})
-	return unclaimed[:minInt(n, len(unclaimed))]
+func (np *nodePool) SampleUnclaimedTasks(ctx context.Context, n int) ([]string, error) {
+	return np.rdb.ZRandMember(ctx, np.key.redisUnclaimedTasksKey(), n, false /* withScores= */).Result()
```
Comment (tylerwilliams) [inline enterprise/server/scheduling/scheduler_server/scheduler_server.go:402]: false /* =withScores */

**Review by tylerwilliams — APPROVED** [review]
LGTM

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go (hunk 1 above)
Comment (vadimberezniker) [reply enterprise/server/scheduling/scheduler_server/scheduler_server.go]: Done.

File: enterprise/server/scheduling/scheduler_server/scheduler_server.go (hunk 2 above)
Comment (vadimberezniker) [reply enterprise/server/scheduling/scheduler_server/scheduler_server.go]: Done.

**Pull request merged**